### ISSUE TITLE
Multiple listeners created for same port

### ISSUE THREAD

**Issue by LoicAG** [Azure/application-gateway-kubernetes-ingress#1151]
This issue is an exact copy of https://github.com/Azure/application-gateway-kubernetes-ingress/issues/966, which was opened by @davejhahn and closed by @akshaysngupta but not yet fully solved.

Attempts to get attention on the closed issue have failed, to I'm re-opening this new issue.

![image](https://user-images.githubusercontent.com/1173317/109640875-4d09fe00-7b51-11eb-97b5-2dc174fad147.png)

**Describe the bug**
I have a configuration where I have multiple domains that should point back to the same backend. With the ingress config I am using it creates two listeners and the same port, and although it appears to work, it shows an error in the portal and on subsequent updates of App GW through Terraform, "This Basic HTTP listener cannot use the same frontend port as an existing listener.

**To Reproduce**
```
spec:
  rules:
  - host: '*.mysite.com'
    http:
      paths:
      - backend:
          serviceName: my-service-name
          servicePort: http
        path: /api/path*
  - host: 'mysite.com'
    http:
      paths:
      - backend:
          serviceName: my-service-name
          servicePort: http
        path: /api/path*
```

This creates two listeners, both on the same port (443 in my case), linked to the front end IP configuration. I'm wondering if it should be multi-site rather than basic but cannot find where that is controlled from the configuration that AGIC uses--I'm thinking it might be automatically determined and that my syntax is wrong but I cannot find any other examples. I read one post about "hostnames" instead of "hostname", but I believe that was internal to AGIC, release notes @ https://github.com/Azure/application-gateway-kubernetes-ingress/blob/master/CHANGELOG/CHANGELOG-1.2.md#v120-rc2

**Ingress Controller details**
```
Name:         ingress-azure-84b5574c46-lk8nf
Namespace:    default
Priority:     0
Node:         aks-default-77266492-vmss000000/10.240.0.4
Start Time:   Mon, 03 Aug 2020 14:55:51 -0500
Labels:       aadpodidbinding=ingress-azure
              app=ingress-azure
              pod-template-hash=84b5574c46
              release=ingress-azure
Annotations:  checksum/config: 34e838db128325bcbaf6bc634e73547ceb97d3acebd83a31f915cd88e3414810
              prometheus.io/port: 8123
              prometheus.io/scrape: true
Status:       Running
IP:           10.240.0.101
IPs:
  IP:           10.240.0.101
Controlled By:  ReplicaSet/ingress-azure-84b5574c46
Containers:
  ingress-azure:
    Container ID:   docker://bf94256cc82672647ab9636bf799316ce91cd13617fafeb555e26cefe7609715
    Image:          mcr.microsoft.com/azure-application-gateway/kubernetes-ingress:1.2.0
    Image ID:       docker-pullable://mcr.microsoft.com/azure-application-gateway/kubernetes-ingress@sha256:de458f962eab0cd2de19d23dfeb9a0e4bc2565a38f8c45cc98a74f3cda8b940c
    Port:           <none>
    Host Port:      <none>
    State:          Running
      Started:      Mon, 03 Aug 2020 14:55:54 -0500
    Ready:          True
    Restart Count:  0
    Liveness:       http-get http://:8123/health/alive delay=15s timeout=1s period=20s #success=1 #failure=3
    Readiness:      http-get http://:8123/health/ready delay=5s timeout=1s period=10s #success=1 #failure=3
    Environment Variables from:
      ingress-azure  ConfigMap  Optional: false
    Environment:
      AZURE_CLOUD_PROVIDER_LOCATION:  /etc/appgw/azure.json
      AGIC_POD_NAME:                  ingress-azure-84b5574c46-lk8nf (v1:metadata.name)
      AGIC_POD_NAMESPACE:             default (v1:metadata.namespace)
    Mounts:
      /etc/appgw/azure.json from azure (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from ingress-azure-token-g89l8 (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  azure:
    Type:          HostPath (bare host directory volume)
    Path:          /etc/kubernetes/azure.json
    HostPathType:  File
  ingress-azure-token-g89l8:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  ingress-azure-token-g89l8
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:          <none>
```

**Ingress Controller Log**
```
I0803 19:55:54.074018       1 auth.go:46] Creating authorizer from Azure Managed Service Identity
I0803 19:55:54.074094       1 httpserver.go:57] Starting API Server on :8123
I0803 19:55:54.951014       1 main.go:177] Ingress Controller will observe the following namespaces:default
I0803 19:55:55.013067       1 context.go:132] k8s context run started
I0803 19:55:55.013098       1 context.go:176] Waiting for initial cache sync
I0803 19:55:55.413244       1 context.go:189] Initial cache sync done
I0803 19:55:55.413265       1 context.go:190] k8s context run finished
I0803 19:55:55.413359       1 worker.go:39] Worker started
I0803 19:55:55.615336       1 mutate_app_gateway.go:177] BEGIN AppGateway deployment
I0803 19:57:26.498775       1 mutate_app_gateway.go:185] Applied generated Application Gateway configuration
I0803 19:57:26.498790       1 mutate_app_gateway.go:200] cache: Updated with latest applied config.
I0803 19:57:26.500565       1 mutate_app_gateway.go:204] END AppGateway deployment
```

**Comment by cedricmenec** [comment]
We have the same problem here, but actually everything works fine. 

AGIC create the listener, and set its "hostnames" field, but Azure Portal did not read this field, it seems to only handle "hostname" field. You can actually check this using Azure CLI:  `az network application-gateway http-listener show`

According to the documentation it seems that multi-sites is a feature that is not yet supported in Azure Portal as mentioned in the documentation ([here](https://docs.microsoft.com/en-us/azure/application-gateway/multiple-site-overview#wildcard-host-names-in-listener-preview:~:text=This%20feature%20is%20currently%20available%20only,until%20it's%20supported%20in%20the%20portal.)).

> This feature is currently available only through Azure PowerShell and Azure CLI. Portal support is coming soon. Please note that since portal support is not fully available, if you are using only the HostNames parameter, the listener will appear as a Basic listener in the portal and the Host name column of the listener list view will not show the host names that are configured. For any changes to a wildcard listener, make sure you use Azure PowerShell or CLI until it's supported in the portal.

Hope that helps.

**Comment by LoicAG** [comment]
Yes, I agree that it's only a display error in the Portal; but in the original issue, it was reported 6 months ago that this display error would "soon" be fixed.

**Comment by cedricmenec** [comment]
It seems that "soon" does not have the same value for everyone...
No, just kidding ! It's just a matter of priority.

However, if an Azure Portal Developer read this issue, we are very interested, here, by the resolution of this issue.
Even if everything, behind the scene, works fine, this display error is annoying and makes our investigation more difficult. Thank you.

**Comment by Expecho** [comment]
Is it really just a display issue? I created a script that associates a rewrite rule set with the routing rules since agic in its infinite wisdom decides to throw them away at every update. But the powershell script fails to update the gateway due to: `Set-AzApplicationGateway : Two Http Listeners of Application Gateway fl-11b6b0ef99b655c6e31b9f4a8008d60f and fl-0e00f020fe85d9dd63eb5e0ef66840cb are using the same Frontend Port`

**Comment by bbreijer** [comment]
Good find @Expecho, maybe it is more than just a display issue after all... Did anyone file a support request for issue this yet?

> .... agic in its infinite wisdom decides to throw them away at every update...

This is mentioned in https://github.com/Azure/application-gateway-kubernetes-ingress/issues/462

**Comment by Expecho** [comment]
@bbreijer I didn't file a support request for this as it is covered by this and other issues. My main concern is that it takes such a long time before it is addressed.

**Comment by LoicAG** [comment]
Yep, it's almost a year since: 
> you should **soon** stop seeing the error message

https://github.com/Azure/application-gateway-kubernetes-ingress/issues/966#issuecomment-668439315

So... happy anniversary? 😒

**Comment by hueseyinkilic** [comment]
> Good find @Expecho, maybe it is more than just a display issue after all... Did anyone file a support request for issue this yet?
> 
> > .... agic in its infinite wisdom decides to throw them away at every update...
> 
> This is mentioned in #462

Yes I have opened a support request in September 2020 about this with Azure and they were looking to deliver a solution EOY 2020 🗡️ 👯‍♂️

**Comment by rowanmoul** [comment]
For anyone finding their way here looking for a way to get AGIC to generate a single listener for a multi-host ingress like the one in the OP, it looks like that is impossible with the current implementation. AGIC loops through each ingress `rule` and creates a separate listener for each one:
https://github.com/Azure/application-gateway-kubernetes-ingress/blob/8bc1e63c3714667bb59f8533fd39b8970e77bd2b/pkg/appgw/ingress_rules.go#L27-L32
There is a workaround! If everything else about your multiple `rules` is the same (paths, etc) then you can use an annotation to add additional hosts, and these are correctly added to the same listener.
`appgw.ingress.kubernetes.io/hostname-extension: "hostname1, hostname2"`
https://azure.github.io/application-gateway-kubernetes-ingress/annotations/#hostname-extension
These additional hostnames are added to ALL `rules` so you probably only want to use this on single-rule ingresses.
Also, make sure your TLS certificates are issued with the necessary wildcards and/or Subject Alternative Names
